src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -42,53 +40,68 @@ internal static T[] ToArray<T>(IEnumerable<T> source, out int length)
-                        const int DefaultCapacity = 4;
-                        int newLength = count == 0 ? DefaultCapacity : count * 2;
-                        if ((uint)newLength > MaxArrayLength)
+                        T[] arr = new T[4];
+                        arr[0] = en.Current;
+                        int count = 1;
+                        
+                        while (en.MoveNext())
                         {
-                            newLength = MaxArrayLength;
-                        }
-                        if (newLength < count + 1)
-                        {
-                            newLength = count + 1;
+                            if (count == arr.Length)
+                            {
+                                // MaxArrayLength is defined in Array.MaxArrayLength and in gchelpers in CoreCLR.
+                                // It represents the maximum number of elements that can be in an array where the size of the
+                                // element is greater than one byte; a separate, slightly larger constant, is used when the
+                                // size of the element is one.
+                                const int MaxArrayLength = 0x7FEFFFFF;
+        
+                                // We attempt to  double the size of the array. Before doubling overflows, it will exceed
+                                // MaxArrayLength, so it is constrained to that size.
+                                // Because a slightly larger constant is used when T is one byte in size, we could then end up
+                                // in a situation where arr.Length is MaxArrayLength or slightly larger, and an overflow does
+                                // happen.
+                                // In this case we increase the array by 2383 each resize as the smallest factor above square
+                                // root of the difference between the maximum one-byte array size (0x7FFFFFC7) and
+                                // MaxArrayLength as a compromise between going straight to 0x7FFFFFC7 (fewer allocations) or
+                                // adding only one element at a time (smaller allocations).
+                                int newLength = count << 1;
+                                if ((uint)newLength > (uint)MaxArrayLength)
+                                {
+                                    if (count < MaxArrayLength)
+                                    {
+                                        newLength = MaxArrayLength;
+                                    }
+                                    else if ((newLength = unchecked(count + 2383)) < 0)
+                                    {
+                                        newLength = int.MaxValue; // Just let the attempt to create an array this size fail.
+                                    }
```

Comment:
defaulting to       newLength = count + 1;   
 seems slightly better here because we do not care much performance-wise (it is a degenerate case and will be slow regardless), but in case someone relied on 1element increases here, we could keep their code running.